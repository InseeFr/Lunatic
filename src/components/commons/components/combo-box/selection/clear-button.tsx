import { useMemo } from 'react';
import classnames from 'classnames';
import { CrossIcon } from '../../../icons';
import { Fab } from '../../fab';
import { customizedComponent } from '../../../../shared/HOC/customizedComponent';

function isDisabled(search?: string) {
	return !search || search.trim().length === 0;
}

function createOnKeyDown(onClick?: () => void) {
	if (typeof onClick !== 'function') {
		return undefined;
	}
	return (e: { key: string }) => {
		if (e.key === 'Enter') {
			onClick();
		}
	};
}

type Props = {
	className?: string;
	search?: string;
	onClick?: () => void;
	editable?: boolean;
};

function ClearButtonComponent({ className, search, onClick, editable }: Props) {
	const onKeyDown = useMemo(() => createOnKeyDown(onClick), [onClick]);
	if (!editable) {
		return null;
	}
	return (
		<Fab
			className={classnames('mini', 'lunatic-combo-box-fab', className)}
			title="delete"
			onClick={onClick}
			disabled={isDisabled(search)}
			onKeyDown={onKeyDown}
		>
			<CrossIcon className="lunatic-combo-box-icon" />
		</Fab>
	);
}

export const ClearButton = customizedComponent(
	'ComboboxClearButton',
	ClearButtonComponent
);
