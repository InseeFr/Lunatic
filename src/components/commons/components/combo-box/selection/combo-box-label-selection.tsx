import classNames from 'classnames';
import { ComboBoxOptionType } from '../combo-box.type';
import createCustomizableLunaticField from '../../../create-customizable-field';

type Props = {
	option?: ComboBoxOptionType | null;
	placeholder?: string;
	search?: string;
	disabled?: boolean;
	readOnly?: boolean;
};

function isPlaceholder(option: Props['option'], search: Props['search']) {
	return !option && (!search || search.length === 0);
}

function getContent(
	option: Props['option'],
	search?: Props['search'],
	placeholder?: string
) {
	if (option) {
		const { id, value, label } = option;
		return label ? `${id || value} - ${label}` : id || value;
	}
	if (search && search.trim().length) {
		return search;
	}
	return placeholder ?? '';
}

function ComboBoxLabelSelection({
	option,
	placeholder,
	search,
	disabled = false,
	readOnly = false,
}: Props) {
	const isPh = isPlaceholder(option, search);
	return (
		<div
			className={classNames('lunatic-combo-box-selected', {
				disabled,
				readOnly,
			})}
		>
			<span
				className={classNames({
					placeholder: isPh,
					selection: !isPh,
				})}
			>
				{getContent(option, search, placeholder)}
			</span>
		</div>
	);
}

export default createCustomizableLunaticField(
	ComboBoxLabelSelection,
	'ComboboxLabelSelection'
);
