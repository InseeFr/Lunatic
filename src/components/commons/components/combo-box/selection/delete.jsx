import React, { useMemo } from 'react';
import classnames from 'classnames';
import { Fab } from '../../fab';
import { CrossIcon } from '../../../icons';

function isDisabled(search) {
	return !search || search.trim().length === 0;
}

function createOnKeyDown(onClick) {
	if (typeof onClick === 'function') {
		return function onKeyDown(e) {
			const { key } = e;
			if (key === 'Enter') {
				onClick();
			}
		};
	}
	return () => null;
}

function Delete({ className, search, onClick, editable }) {
	const onKeyDown = useMemo(() => createOnKeyDown(onClick), [onClick]);
	if (editable) {
		return (
			<Fab
				className={classnames('mini', 'lunatic-combo-box-fab', className)}
				title="delete"
				onClick={onClick}
				disabled={isDisabled(search)}
				onKeyDown={onKeyDown}
			>
				<CrossIcon className="lunatic-combo-box-icon" />
			</Fab>
		);
	}
	return null;
}

export default Delete;
