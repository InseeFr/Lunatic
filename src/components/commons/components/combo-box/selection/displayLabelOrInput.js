import React from 'react';

function displayLabelOrInput(Field) {
	const Memoized = React.memo(Field);
	const { name } = Field;

	return function LabelOrInput(props) {
		const { editable, expended } = props;

		const displayLabel = !editable || !expended;

		if (
			(name === 'Input' && !displayLabel) ||
			(name === 'Label' && displayLabel)
		) {
			return <Memoized {...props} />;
		}

		return null;
	};
}

export default displayLabelOrInput;
