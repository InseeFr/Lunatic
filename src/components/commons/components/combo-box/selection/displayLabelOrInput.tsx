import { FunctionComponent, memo } from 'react';

export function displayLabelOrInput<T extends Record<string, unknown>>(
	Field: FunctionComponent<T>,
	name: string
) {
	const Memoized = memo<T>(Field);

	return function LabelOrInput(props: T) {
		const { editable, expanded } = props;

		const displayLabel = !editable || !expanded;

		if (
			(name === 'Input' && !displayLabel) ||
			(name === 'LabelSelection' && displayLabel)
		) {
			return <Memoized {...props} />;
		}

		return null;
	};
}
