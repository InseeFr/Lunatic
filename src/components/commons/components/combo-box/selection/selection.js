import React, { useCallback } from 'react';
import LabelSelection from './label-selection';
import Input from './input';
import SelectionContainer from './selection-container';

function Selection({
	labelRenderer,
	placeholder,
	search,
	expended,
	htmlFor,
	disabled,
	focused,
	onChange,
	selectedIndex,
	options,
	editable,
	labelledBy,
	id,
}) {
	const onChangeEx = useCallback(
		function (e) {
			onChange(e.target.value);
		},
		[onChange]
	);

	return (
		<SelectionContainer
			id={id}
			aria-controls={'todo'}
			expanded={expended}
			ariaOwns={`${id}-list`}
		>
			<LabelSelection
				id={htmlFor}
				labelRenderer={labelRenderer}
				placeholder={placeholder}
				editable={editable}
				selectedIndex={selectedIndex}
				options={options}
				search={search}
				disabled={disabled}
				labelledBy={labelledBy}
				expended={expended}
			/>

			<Input
				id={htmlFor}
				className="lunatic-combo-box-input"
				onChange={onChangeEx}
				value={search}
				placeholder={placeholder}
				disabled={disabled}
				focused={focused}
				editable={editable}
				expended={expended}
			/>
		</SelectionContainer>
	);
}

export default Selection;
