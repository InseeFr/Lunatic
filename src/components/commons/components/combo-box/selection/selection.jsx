import React, { useCallback } from 'react';
import LabelSelection from './label-selection';
import Input from './input';
import SelectionContainer from './selection-container';

function Selection({
	labelRenderer,
	placeholder,
	search,
	expended,
	disabled,
	focused,
	onChange,
	selectedIndex,
	options,
	editable,
	labelId,
	id,
}) {
	const onChangeEx = useCallback(
		function (e) {
			onChange(e.target.value);
		},
		[onChange]
	);

	return (
		<SelectionContainer
			id={id}
			labelId={labelId}
			expanded={expended}
			ariaOwns={`${id}-list`}
		>
			<LabelSelection
				labelRenderer={labelRenderer}
				placeholder={placeholder}
				editable={editable}
				selectedIndex={selectedIndex}
				options={options}
				search={search}
				disabled={disabled}
				expended={expended}
			/>
			<Input
				id={`combobox-input-${id}`}
				className="lunatic-combo-box-input"
				onChange={onChangeEx}
				value={search}
				placeholder={placeholder}
				disabled={disabled}
				focused={focused}
				editable={editable}
				expended={expended}
			/>
		</SelectionContainer>
	);
}

export default Selection;
