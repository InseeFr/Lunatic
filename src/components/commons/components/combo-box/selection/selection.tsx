import { ChangeEventHandler, useCallback } from 'react';
import { LabelSelectionProps } from './label-selection';
import SelectionContainer from './selection-container';
import { LabelOrInput } from './LabelOrInput';

export type SelectionProps = {
	expanded?: boolean;
	focused?: boolean;
	onChange?: (s: string | null) => void;
	editable?: boolean;
	labelId?: string;
	id?: string;
	className?: string;
} & LabelSelectionProps;

export function Selection({
	labelRenderer,
	placeholder,
	search,
	expanded,
	disabled,
	focused,
	onChange,
	selectedIndex,
	options,
	editable,
	labelId,
	id,
	className,
}: SelectionProps) {
	const onChangeEx = useCallback<ChangeEventHandler<HTMLInputElement>>(
		(e) => {
			onChange?.(e.target.value);
		},
		[onChange]
	);

	return (
		<SelectionContainer
			id={id}
			labelId={labelId}
			expanded={expanded}
			aria-owns={`${id}-list`}
			className={className}
		>
			<LabelOrInput
				labelRenderer={labelRenderer}
				placeholder={placeholder}
				selectedIndex={selectedIndex}
				options={options}
				search={search}
				disabled={disabled}
				onChange={onChangeEx}
				value={search}
				focused={focused}
				editable={editable}
				expanded={expanded}
			/>
		</SelectionContainer>
	);
}
