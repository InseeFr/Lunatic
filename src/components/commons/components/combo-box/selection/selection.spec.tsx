import { render } from '@testing-library/react';
import { Selection } from './selection';
import { describe, expect, it } from 'vitest';

const options = [
	{ label: 'Apple', value: 'apple' },
	{ label: 'Orange', value: 'orange' },
	{ label: 'Banana', value: 'banana' },
];

describe('Selection component', () => {
	it('should render correctly with default props', () => {
		const { container } = render(<Selection options={options} />);

		expect(container.firstChild).toMatchSnapshot();
	});

	it('should render correctly when it is editable', () => {
		const { container } = render(
			<Selection options={options} editable={true} />
		);

		expect(container.firstChild).toMatchSnapshot();
	});

	it('should render correctly when it is expanded', () => {
		const { container } = render(
			<Selection options={options} expanded={true} />
		);

		expect(container.firstChild).toMatchSnapshot();
	});

	it('should render correctly when it is disabled', () => {
		const { container } = render(
			<Selection options={options} disabled={true} placeholder="hello!" />
		);

		expect(container.firstChild).toMatchSnapshot();
	});

	it('should render the selected option when there is a selection', () => {
		const { getByText } = render(
			<Selection options={options} selectedIndex={0} expanded={false} />
		);

		expect(getByText('apple - Apple')).toBeInTheDocument();
	});

	it('should not render the input when it is not editable', () => {
		const { queryByPlaceholderText } = render(
			<Selection
				options={options}
				placeholder="Search fruits"
				editable={false}
			/>
		);

		expect(queryByPlaceholderText('Search fruits')).not.toBeInTheDocument();
	});
});
