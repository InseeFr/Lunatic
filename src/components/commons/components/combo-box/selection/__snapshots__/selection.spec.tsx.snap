// Vitest Snapshot v1

exports[`Selection component > should render correctly when it is disabled 1`] = `
<div
  aria-autocomplete="list"
  aria-controls="todo"
  aria-haspopup="listbox"
  aria-owns="idUndefined"
  class="lunatic-combo-box-selection"
  id="idUndefined"
  role="combobox"
>
  <div
    class="lunatic-combo-box-selected disabled"
  >
    <span
      class="placeholder"
    >
      hello!
    </span>
  </div>
</div>
`;

exports[`Selection component > should render correctly when it is editable 1`] = `
<div
  aria-autocomplete="list"
  aria-controls="todo"
  aria-haspopup="listbox"
  aria-owns="idUndefined"
  class="lunatic-combo-box-selection"
  id="idUndefined"
  role="combobox"
>
  <div
    class="lunatic-combo-box-selected"
  >
    <span
      class="placeholder"
    />
  </div>
</div>
`;

exports[`Selection component > should render correctly when it is expanded 1`] = `
<div
  aria-autocomplete="list"
  aria-controls="todo"
  aria-expanded="true"
  aria-haspopup="listbox"
  aria-owns="idUndefined"
  class="lunatic-combo-box-selection"
  id="idUndefined"
  role="combobox"
>
  <div
    class="lunatic-combo-box-selected"
  >
    <span
      class="placeholder"
    />
  </div>
</div>
`;

exports[`Selection component > should render correctly with default props 1`] = `
<div
  aria-autocomplete="list"
  aria-controls="todo"
  aria-haspopup="listbox"
  aria-owns="idUndefined"
  class="lunatic-combo-box-selection"
  id="idUndefined"
  role="combobox"
>
  <div
    class="lunatic-combo-box-selected"
  >
    <span
      class="placeholder"
    />
  </div>
</div>
`;
